CredentialProvider.Microsoft/Util/FeedEndpointCredentialsParser.cs

```diff
@@ -0,0 +1,171 @@
+
+public static class FeedEndpointCredentialsParser
+{
+    private static readonly JsonSerializerOptions options = new JsonSerializerOptions
+    {
+        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+        PropertyNameCaseInsensitive = true,
+    };
+
+    public static Dictionary<string, EndpointCredentials> ParseFeedEndpointsJsonToDictionary(ILogger logger)
+    {
+        string feedEndPointsJson = Environment.GetEnvironmentVariable(EnvUtil.EndpointCredentials);
+        if (string.IsNullOrWhiteSpace(feedEndPointsJson))
+        {
+            return new Dictionary<string, EndpointCredentials>(StringComparer.OrdinalIgnoreCase);
+        }
+
+        try
+        {
+            logger.Verbose(Resources.ParsingJson);
+            Dictionary<string, EndpointCredentials> credsResult = new Dictionary<string, EndpointCredentials>(StringComparer.OrdinalIgnoreCase);
+            EndpointCredentialsContainer endpointCredentials = System.Text.Json.JsonSerializer.Deserialize<EndpointCredentialsContainer>(feedEndPointsJson, options);
+            if (endpointCredentials == null)
+            {
+                logger.Verbose(Resources.NoEndpointsFound);
+                return credsResult;
+            }
+
+            foreach (var credentials in endpointCredentials.EndpointCredentials)
+            {
+                if (credentials == null)
+                {
+                    logger.Verbose(Resources.EndpointParseFailure);
+                    break;
+                }
+
+                if (credentials.ClientId == null)
+                {
+                    logger.Verbose(Resources.EndpointParseFailure);
+                    break;
+                }
+
+                if(credentials.CertificateSubjectName != null && credentials.CertificateFilePath != null)
+                {
+                    logger.Verbose(Resources.EndpointParseFailure);
+                    break;
+                }
+
+                if (!Uri.TryCreate(credentials.Endpoint, UriKind.Absolute, out var endpointUri))
+                {
+                    logger.Verbose(Resources.EndpointParseFailure);
+                    break;
+                }
+
+                var urlEncodedEndpoint = endpointUri.AbsoluteUri;
+                if (!credsResult.ContainsKey(urlEncodedEndpoint))
+                {
+                    credsResult.Add(urlEncodedEndpoint, credentials);
+                }
+            }
+
+            return credsResult;
+        }
+        catch (Exception e)
+        {
+            logger.Verbose(string.Format(Resources.VstsBuildTaskExternalCredentialCredentialProviderError, e));
+            return new Dictionary<string, EndpointCredentials>(StringComparer.OrdinalIgnoreCase); ;
+        }
+    }
+
+
+    public static Dictionary<string, ExternalEndpointCredentials> ParseExternalFeedEndpointsJsonToDictionary(ILogger logger)
+    {
+        string feedEndPointsJson = Environment.GetEnvironmentVariable(EnvUtil.BuildTaskExternalEndpoints);
```

Comment:
endpoint not endPoints... endpoint is 1 word, not 2 (though end and point are also words, but that's irrelevant)

Reply:
done!